staging/vhost-device-gpu/src/virtio_gpu.rs

```diff
@@ -0,0 +1,1055 @@
+    ) -> VirtioGpuResult {
+        let buf = buf.map(|vs| {
+            IoSliceMut::new(
+                // SAFETY: trivially safe
+                unsafe { std::slice::from_raw_parts_mut(vs.ptr_guard_mut().as_ptr(), vs.len()) },
+            )
+        });
+        self.rutabaga
+            .transfer_read(ctx_id, resource_id, transfer, buf)?;
+        Ok(OkNoData)
+    }
+
+    fn attach_backing(
+        &mut self,
+        resource_id: u32,
+        mem: &GuestMemoryMmap,
+        vecs: Vec<(GuestAddress, usize)>,
+    ) -> VirtioGpuResult {
+        let rutabaga_iovecs = sglist_to_rutabaga_iovecs(&vecs[..], mem).map_err(|_| ErrUnspec)?;
+        self.rutabaga.attach_backing(resource_id, rutabaga_iovecs)?;
+        Ok(OkNoData)
+    }
+
+    fn detach_backing(&mut self, resource_id: u32) -> VirtioGpuResult {
+        self.rutabaga.detach_backing(resource_id)?;
+        Ok(OkNoData)
+    }
+
+    fn update_cursor(
+        &mut self,
+        resource_id: u32,
+        gpu_backend: &mut GpuBackend,
+        cursor_pos: VhostUserGpuCursorPos,
+        hot_x: u32,
+        hot_y: u32,
+    ) -> VirtioGpuResult {
+        let mut data = Box::new([0; 4 * 64 * 64]);
+        let cursor_rect = Rectangle {
+            x: 0,
+            y: 0,
+            width: 64,
+            height: 64,
+        };
+
+        self.read_2d_resource(resource_id, &cursor_rect, &mut data[..])
+            .map_err(|e| {
+                error!("Failed to read resource of cursor: {e}");
+                ErrUnspec
+            })?;
+
+        let cursor_update = VhostUserGpuCursorUpdate {
+            pos: cursor_pos,
+            hot_x,
+            hot_y,
+        };
+
+        gpu_backend
+            .cursor_update(&cursor_update, &data)
+            .map_err(|e| {
+                error!("Failed to update cursor pos from frontend: {}", e);
+                ErrUnspec
+            })?;
+
+        Ok(OkNoData)
+    }
+
+    fn move_cursor(
+        &mut self,
+        resource_id: u32,
+        gpu_backend: &mut GpuBackend,
+        cursor: VhostUserGpuCursorPos,
+    ) -> VirtioGpuResult {
+        if resource_id == 0 {
+            gpu_backend.cursor_pos_hide(&cursor).map_err(|e| {
+                error!("Failed to set cursor pos from frontend: {}", e);
+                ErrUnspec
+            })?;
+        } else {
+            gpu_backend.cursor_pos(&cursor).map_err(|e| {
+                error!("Failed to set cursor pos from frontend: {}", e);
+                ErrUnspec
+            })?;
+        }
+
+        Ok(OkNoData)
+    }
+
+    fn resource_assign_uuid(&self, resource_id: u32) -> VirtioGpuResult {
+        if !self.resources.contains_key(&resource_id) {
+            return Err(ErrInvalidResourceId);
+        }
+
+        // TODO(stevensd): use real uuids once the virtio wayland protocol is updated to
+        // handle more than 32 bits. For now, the virtwl driver knows that the uuid is
+        // actually just the resource id.
```

Comment:
I thought `virtio-wayland` was discontinued in favor of `virtio-gpu` and `virtio-video`, no?

Reply:
In any case it's a crosvm thing, not a `VIRTIO` spec thing

Reply:
yes this bit was ported from crosvm, hence the comment.